src/Storages/StoragePostgreSQL.cpp

```diff
@@ -0,0 +1,177 @@
+{
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(columns_);
+    storage_metadata.setConstraints(constraints_);
+    setInMemoryMetadata(storage_metadata);
+}
+
+
+Pipe StoragePostgreSQL::read(
+    const Names & column_names_,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info_,
+    const Context & context_,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size_,
+    unsigned)
+{
+    metadata_snapshot->check(column_names_, getVirtuals(), getStorageID());
+
+    String query = transformQueryForExternalDatabase(
+        query_info_, metadata_snapshot->getColumns().getOrdinary(),
+        IdentifierQuotingStyle::DoubleQuotes, "", remote_table_name, context_);
+
+    Block sample_block;
+    for (const String & column_name : column_names_)
+    {
+        auto column_data = metadata_snapshot->getColumns().getPhysical(column_name);
+        WhichDataType which(column_data.type);
+        if (which.isEnum())
+            column_data.type = std::make_shared<DataTypeString>();
+        sample_block.insert({ column_data.type, column_data.name });
+    }
+
+    return Pipe(std::make_shared<SourceFromInputStream>(
+            std::make_shared<PostgreSQLBlockInputStream>(connection->conn(), query, sample_block, max_block_size_)));
+}
+
+
+BlockOutputStreamPtr StoragePostgreSQL::write(
+        const ASTPtr & /*query*/, const StorageMetadataPtr & metadata_snapshot, const Context & /* context */)
+{
+    return std::make_shared<PostgreSQLBlockOutputStream>(metadata_snapshot, connection->conn(), remote_table_name);
+}
+
+
+void PostgreSQLBlockOutputStream::writePrefix()
+{
+    work = std::make_unique<pqxx::work>(*connection);
+}
+
+
+void PostgreSQLBlockOutputStream::write(const Block & block)
+{
+    const auto columns = block.getColumns();
+    const size_t num_rows = block.rows(), num_cols = block.columns();
+    const auto data_types = block.getDataTypes();
+
+    if (!stream_inserter)
+        stream_inserter = std::make_unique<pqxx::stream_to>(*work, remote_table_name, block.getNames());
+
+    /// std::optional lets libpqxx to know if value is NULL
+    std::vector<std::optional<std::string>> row(num_cols);
+
+    for (const auto i : ext::range(0, num_rows))
+    {
+        for (const auto j : ext::range(0, num_cols))
+        {
+            if (columns[j]->isNullAt(i))
+            {
+                row[j] = std::nullopt;
+            }
+            else
+            {
+                WriteBufferFromOwnString ostr;
+                data_types[j]->serializeAsText(*columns[j], i, ostr, FormatSettings{});
+                row[j] = std::optional<std::string>(ostr.str());
+
+                if (isArray(data_types[j]))
+                {
+                    char r;
+                    std::replace_if(row[j]->begin(), row[j]->end(), [&](char c)
+                    {
+                        return ((c == '[') && (r = '{')) || ((c == ']') && (r = '}'));
```

Comment:
This is questionable.
Suppose we have an array
```
['Hel[lo', 'Wor]ld']
```

And how do we handle multidimensional arrays?

Reply:
> And how do we handle multidimensional arrays?

It was handled ok because in this line
`data_types[j]->serializeAsText(*columns[j], i, ostr, FormatSettings{});` multidimensional arrays were also converted, but the problem was that it enclosed arrays in `[]` and for Postgres it needs to be `{}`. 

> This is questionable.

Yes, I rewrote this part completely. Is it better now?